### ISSUE TITLE
`unnecessary_cast` false positive when cast affects inferred type

### ISSUE THREAD

**Issue by jamesderlin** [dart-lang/linter#4276]
**Describe the issue**

`x as T` seems to trigger an `unnecessary_cast` lint if the static type of `x` is statically known to be a subtype of `T`.  Although such a cast superficially seems pointless, there can be cases where the cast makes a difference.  For example, consider:

```dart
var record = (1, 2, 3);
```
The inferred type of `record` is `(int, int, int)`.  If I want one of the members to be a `num`, I could explicitly specify the `Record` type:
```dart
(int, num, int) record = (1, 2, 3);
```
but that becomes awkward, tedious, and potentially error-prone if the `Record` has many members.  I alternatively could do:
```dart
num x = 2;
var record = (1, x, 3);
```
which I suppose is okay, but it's more verbose and requires naming more things.  I'd rather do:
```dart
var record = (1, 2 as num, 3);
```
The last version, however, triggers the `unnecessary_cast` lint (using DartPad based on Flutter 3.10.0-4.0.pre Dart SDK 3.0.0-431.0.dev).

**To Reproduce**

See above.

**Expected behavior**

Ideally the `unnecessary_cast` lint would trigger if and only if removing the cast has no effect.

I admit that this sort of situation probably isn't likely to occur much in practice.
